### DISCUSSION TITLE
How major GC handles orphan segments and blocks?

### DISCUSSION THREAD

**Discussion by zwang28** [databendlabs/databend#9149] [Q&A] [answered]
[As mentioned](https://github.com/datafuselabs/databend/blob/0a902763ec1b37a16344050f7eb4217bd464dd68/src/query/storages/fuse/fuse/src/operations/commit.rs#L675) if deletion operation for blocks and segments failed, they will be collected by major GC. IIUC major GC refers to [do_gc](https://github.com/datafuselabs/databend/blob/0a902763ec1b37a16344050f7eb4217bd464dd68/src/query/storages/fuse/fuse/src/operations/gc.rs#L41). But I cannot figure out how segments and blocks that is not tracked by any snapshot files can be deleted by `do_gc`, or I'm missing something? Thanks.

**Comment by dantengsky** [comment, marked as answer]
> IIUC major GC refers to [do_gc](https://github.com/datafuselabs/databend/blob/0a902763ec1b37a16344050f7eb4217bd464dd68/src/query/storages/fuse/fuse/src/operations/gc.rs#L41)

the "major GC" is still under development. 

> But I cannot figure out how segments and blocks that is not tracked by any snapshot files...

for the first iteration, the "major GC' will be implemented in a straightforward way, listing all the data files we are interested in,  if they are not referenced by the current snapshot (or the GC root specified), and out of  the retention period,  they should be collected.
